### ISSUE TITLE
Build flow jobs: options for simple jobs array, or a full buildflow config

### ISSUE THREAD

**Issue by marcesher** [cfpb/jenkins-automation#17]
@imuchnik 
This stems from a conversation I had with Vic this morning about OaH's build flow and how it's not a straightforward list of jobs but instead has a parallel block.

My comment to Vic was to keep this as simple as possible, not get clever or fancy, not reinvent any wheels, and not fall into the trap of the "wrapper" being more complicated than the thing it's wrapping.

I recommended having our build flow builder offer 2 possibilities for creating a flow job: 1) an array of job names; 2) a closure that contains the entire build flow

For #2, I should've said string, based on my now-corrected understanding of buildflow job-dsl (https://jenkinsci.github.io/job-dsl-plugin/#method/javaposse.jobdsl.dsl.DslFactory.buildFlowJob)

Basically, I should be able to do something like:

`new BuildFlowBuilder(blahblah, jobs: ["job1","job2"])`

Or

```
new BuildFlowBuilder(blahblah, jobConfig: '''
build("job1")
parallel {
  build("job2")
  build("job3")
}
build("job4")
'''
```

Or some such thing. But nothing more complicated than that. Basically, make easy things easy (option #1), and hard things possible (option #2), but without making option #2 so onerous (by being clever or fancy) that people rip out their hair.

Thoughts?

**Comment by marcesher** [comment]
Recreating our conversation from email in an issue, and then we can continue the discussion here. 

---

Irina says:

Firstly, I want to let you know that I really enjoy "blah blah"s in a technical context.  Consider further enhancing it with "yada yada yada". 

There is sort of tension between making builders 'lean, but very useful' and 'doing everything, but too clever and annoying". Out initial decision was to make them really simple to decrease verbosity and have more concise and terse dsl for common tasks, while not limiting an ability to extend easily. So following Marc's example a more complicated flow job would look like:

```
// Create a basic flow job
def projectFlowJob= new FlowJobBuilder(
        name: 'Master Flow',
        description: This kicks off put build"',
        jobs:['job1', 'job2']
).build(this);

//add or override values you need using with closure projectFlowJob.with{
  logRotator{
      numToKeep(365)
  }
}
```

 If we would see a repeating pattern in override logic, we'd pull into the builder with a sensible default and/or  make it optional.  

Does that align well with you vision? Thoughts?

--im

---

Marc says:

I think we're talking about two different things.

When I talk about a complicated flow, I'm not talking about additional configuration outside the flow, which is the example you provide and which I heartily agree with.

What I'm talking about is the build flow config itself. Right now, the FlowBuilder only supports an array of jobs, which is necessary but not sufficient. The single recommendation I'm making here is that it also support a string -- perhaps in a "flowConfig" parameter or some such thing, so that I can completely control the flow DSL rather than passing in an array of jobs. In this way, if I have needs other than a simple list of jobs, I can pass in _exactly_ the string that I'd put in the flow DSL block were I configuring this job in the GUI, and the FlowBuilder will simply pass that through.

Check this out:  https://[GHE]/marcesher/qu-jenkins-automation/blob/master/jobs/BuildFlows.groovy

The difference between our current reality, and what I'm proposing, is simply that instead of me having to do this:

```
configure { node ->
        (node / dsl).value = '''
build("qu-dev-build-master")
build("qu-dev-deploy-master")
parallel (
  { build("qu-dev-record-deploy") },
  { build("qu-dev-postdeploy-smoketests") },
  { build("qu-dev-bdd-security") }
).....
```

I'd do this:

```
new FlowJobBuilder(
        name: 'qu-dev-build-flow',
        description: 'Build flow for Qu deployment',
    flowConfig: '''
build("qu-dev-build-master")
build("qu-dev-deploy-master")
parallel (
  { build("qu-dev-record-deploy") },
  { build("qu-dev-postdeploy-smoketests") },
  { build("qu-dev-bdd-security") }
)...
'''
```

---

Irina says:

Ah, got it.  You want to be able to pass in either a list of jobs or a  config. Is there anything a reason to pass a string, rather then a closure?

Something like this:

```
def config =dsl{
'''
    build("qu-dev-build-master")
    build("qu-dev-deploy-master")
    parallel (
        { build("qu-dev-record-deploy") },
        { build("qu-dev-postdeploy-smoketests") },
         { build("qu-dev-bdd-security") }
    )
    retry(2) {
     build("qu-dev-load-test-queries-json")
     build("qu-dev-load-test-aggregations-json")
    ''')
}

new FlowJobBuilder(
        name: 'qu-dev-build-flow',
        description: 'Build flow for Qu deployment',
        jobs:${config}
).build(this)
```

---

Marc says:

Because passing a closure adds no more benefit over passing a string, I'd say. I can't think of any benefit of using

Dsl {...} vs ''' '''

What am I missing?

---

Irina says:

In this case, may be not much. 
But in general closures over strings:  
    1. Closures can be evaluated
    2. The builder will need to convert to closure, so might as well pass one?

**Comment by marcesher** [comment]
@imuchnik So to your last point, re converting to closure, I'd say that's really not our concern, and that what's more important is api compatibility.

From the job-dsl api:

```
buildFlow('''
        build("job1")
    ''')
```

So what I'm proposing is that we simply add an option to our builder that exactly allows anyone to simply use the API directly, basically. They don't need to know "oh, you need to wrap it in this dsl{} block". Our code remains simple because it passes the build flow verbatim into the `buildFlow()` config. This lets someone like me, with a more complicated flow (explained above) to take advantage of the flowbuilder (and, really, the BaseJobBuilder) while not resorting to weird `(blah).value = '''...'''` stuff and with simply copying/pasting my existing build flow DSL from my point/click jenkins job into my new automation script and then I'm done

**Comment by imuchnik** [comment]
@marcesher  Fair point. Delegates is another reason to use closures instead of strings.  But you are being fairly explicit in the expectation, so I rest my case.

**Comment by marcesher** [comment]
Also, @imuchnik, re:

> > > Firstly, I want to let you know that I really enjoy "blah blah"s in a technical context. Consider further enhancing it with "yada yada yada".

Noted! :laughing:

**Comment by OrlandoSoto** [comment]
great thread! I'm glad to see we can keep the parallel job feature in our build flows. This feature has been very useful in reducing the execution time of a build flow by running browser tests and security tests in parallel, since those do not interfere with each other

**Comment by marcesher** [comment]
@OrlandoSoto We will ensure that we have the entire functionality of buildflow available in this builder.

**Comment by marcesher** [comment]
In fact, rule #1 for us in all these builders -- whose purpose is to simplify life a bit -- is probably "do no harm", or perhaps "don't make anything harder than it already is". Simplify hard things, and ensure easy things aren't made less easy

**Comment by OrlandoSoto** [comment]
simplicity is beautiful :)

**Comment by imuchnik** [comment]
Guiding principles in no particular order:
1. [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it)
2. [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)
3. [KISS](https://en.wikipedia.org/wiki/KISS_principle)
